### PR TITLE
[2/n] Instrument sensors page for pageload + move workspace query to indexedb + speed up workspace pages

### PR DESCRIPTION
**Pull request by salazarm**
## Summary & Motivation

- Instrument the sensors page by adding `useBlockTraceOnQueryResult` where approriate
- Remove artificial query waterfall caused by WorkspaceRoot delaying children from rendering until the workspace query is done loading
  - I made the sensor page use the result from the workspace query if its available, otherwise it makes its own query for now
  - Improved a large workspace loading time A LOT (loading a sensor was artificially waiting 40+ seconds, now we don't wait so thats -40 seconds in all cases).
 

## How I Tested These Changes

Went through all the workspace pages and made sure they loaded fine.

### CODE REVIEW

File: js_modules/dagster-ui/packages/ui-core/src/assets/auto-materialization/AutomaterializationTickDetailDialog.tsx [hunk 1]
```diff
@@ -202,6 +203,9 @@ const AssetDetailRow = ({
       },
     },
   );
+  const {data} = queryResult;
+  useBlockTraceOnQueryResult(queryResult, 'AssetGroupAndLocationQuery');
```
Comment (bengotow) [inline js_modules/dagster-ui/packages/ui-core/src/assets/auto-materialization/AutomaterializationTickDetailDialog.tsx:207]: I wonder if there's a way to read the query name out of the queryResult / apollo context here so that we don't need to put the name into a string each time... I feel like these could fall out of sync

Comment (bengotow) [reply js_modules/dagster-ui/packages/ui-core/src/assets/auto-materialization/AutomaterializationTickDetailDialog.tsx:207]: edit: hrmm i guess not, there's no reference to the `query` in the queryResult. Lame!

File: js_modules/dagster-ui/packages/ui-core/src/assets/auto-materialization/SensorAutomaterializationEvaluationHistoryTable.tsx [hunk 2]
```diff
@@ -83,6 +84,8 @@ export const SensorAutomaterializationEvaluationHistoryTable = ({
     pageSize: PAGE_SIZE,
   });
 
+  useBlockTraceOnQueryResult(queryResult, 'AssetSensorTicksQuery');
```
Comment (bengotow) [inline js_modules/dagster-ui/packages/ui-core/src/assets/auto-materialization/SensorAutomaterializationEvaluationHistoryTable.tsx:87]: Might be worth driving through all the pages and putting this hook in most places where we are using `useQueryRefreshAtInterval`. I think generally the one that refreshes is the main data query for the page

File: js_modules/dagster-ui/packages/ui-core/src/overview/OverviewAssetsRoot.tsx [hunk 3]
```diff
@@ -44,7 +44,7 @@ export const OverviewAssetsRoot = ({Header, TabButton}: Props) => {
   useDocumentTitle('Overview | Assets');
 
   const {assets, query, error, loading} = useAllAssets();
-  const refreshState = useRefreshAtInterval({
+  const refreshState = useRefreshAtInterval<any>({
```
Comment (bengotow) [inline js_modules/dagster-ui/packages/ui-core/src/overview/OverviewAssetsRoot.tsx:47]: Was this necessary? Doesn't seem related to other stuff in this PR

File: js_modules/dagster-ui/packages/ui-core/src/workspace/WorkspaceContext.tsx [hunk 4]
```diff
@@ -178,6 +183,7 @@ export const ROOT_WORKSPACE_QUERY = gql`
   ${PYTHON_ERROR_FRAGMENT}
   ${REPOSITORY_INFO_FRAGMENT}
   ${SENSOR_SWITCH_FRAGMENT}
+  ${BASIC_INSTIGATION_STATE_FRAGMENT}
```
Comment (bengotow) [inline js_modules/dagster-ui/packages/ui-core/src/workspace/WorkspaceContext.tsx:186]: I'm not sure we should be adding more stuff to the root workspace query, since we're hoping to get rid of it? It seems like this PR is a nice improvement but moves in the opposite direction (more root workspace, with stronger caching)

File: js_modules/dagster-ui/packages/ui-core/src/workspace/WorkspaceContext.tsx [hunk 5]
```diff
@@ -186,9 +192,17 @@ export const ROOT_WORKSPACE_QUERY = gql`
...
  */
 const useWorkspaceState = (): WorkspaceState => {
-  const {data, loading, refetch} = useQuery<RootWorkspaceQuery, RootWorkspaceQueryVariables>(
-    ROOT_WORKSPACE_QUERY,
-  );
+  const {
+    data,
+    loading,
+    fetch: refetch,
+  } = useIndexedDBCachedQuery<RootWorkspaceQuery, RootWorkspaceQueryVariables>({
```
Comment (bengotow) [inline js_modules/dagster-ui/packages/ui-core/src/workspace/WorkspaceContext.tsx:199]: This is great!  I'm a big fan but we should verify this doesn't break anything in scenarios where the root workspace clearing on a full page load might have been good.

eg: switching orgs in cloud --  since the cache key is constant, maybe when you open one org you will now see the other org's root workspace info for a minute?

File: js_modules/dagster-ui/packages/ui-core/src/overview/OverviewAssetsRoot.tsx (hunk 3 above)
Comment (salazarm) [reply js_modules/dagster-ui/packages/ui-core/src/overview/OverviewAssetsRoot.tsx:47]: yeah it is, switching Workspace to useIndexedDBCachedQuery revealed some weirdness in the typing

File: js_modules/dagster-ui/packages/ui-core/src/workspace/WorkspaceContext.tsx (hunk 4 above)
Comment (salazarm) [reply js_modules/dagster-ui/packages/ui-core/src/workspace/WorkspaceContext.tsx:186]: We're thinking of actually going completely in that other direction. Full root workspace query with indexedb caching, relying on the deployment status query to figure out when a workspace is stale or not and just automagically managing it for the app rather than having you needing to fetch information.

The reason I added these fields was to make the cached result usable by the sensors page in order to immediately use this data if available.

File: js_modules/dagster-ui/packages/ui-core/src/workspace/WorkspaceContext.tsx (hunk 5 above)
Comment (salazarm) [reply js_modules/dagster-ui/packages/ui-core/src/workspace/WorkspaceContext.tsx:199]: Our orgs all use subdomains which should isolate them, but it is actually possible to access ur org without a subdomain so yeah I'll add the org name as part of the key for now

File: js_modules/dagster-ui/packages/ui-core/src/overview/OverviewSensors.tsx [hunk 6]
```diff
@@ -94,7 +102,13 @@ export const OverviewSensors = () => {
...
     },
   );
-  const {data, loading} = queryResultOverview;
+  const {data: currentData, loading} = queryResultOverview;
+  const data =
+    currentData ??
+    (cachedData?.workspaceOrError.__typename === 'Workspace'
+      ? (cachedData as Extract<typeof cachedData, {workspaceOrError: {__typename: 'Workspace'}}>)
+      : null);
+  useBlockTraceOnQueryResult(queryResultOverview, 'OverviewSensorsQuery');
```
Comment (alangenfeld) [inline js_modules/dagster-ui/packages/ui-core/src/overview/OverviewSensors.tsx:111]: why can't we just drop the need for this query entirely?

Comment (salazarm) [reply js_modules/dagster-ui/packages/ui-core/src/overview/OverviewSensors.tsx:111]: That will be part of a much bigger refactor were we kill all of the sub queries. FWIW this query does occur faster than the root workspace query, so technically we'd be regressing perf (at least on an uncached load) initially

Comment (alangenfeld) [reply js_modules/dagster-ui/packages/ui-core/src/overview/OverviewSensors.tsx:111]: got it - can we leave some blob of comments articulating the current state and expected future state

File: js_modules/dagster-ui/packages/ui-core/src/workspace/WorkspaceRoot.tsx [hunk 7]
```diff
@@ -45,12 +47,6 @@ const RepoRouteContainer = () => {
     );
   }
 
-  const {loading} = workspaceState;
-
-  if (loading) {
-    return <div />;
-  }
```
Comment (alangenfeld) [inline js_modules/dagster-ui/packages/ui-core/src/workspace/WorkspaceRoot.tsx:52]: does this change effect other pages that we should be concerned about?

File: js_modules/dagster-ui/packages/ui-core/src/overview/OverviewAssetsRoot.tsx (hunk 3 above)
Comment (alangenfeld) [reply js_modules/dagster-ui/packages/ui-core/src/overview/OverviewAssetsRoot.tsx:47]: > Managing a stack is hard

graphite cli is worth trying if you haven't

File: js_modules/dagster-ui/packages/ui-core/src/workspace/WorkspaceRoot.tsx (hunk 7 above)
Comment (salazarm) [reply js_modules/dagster-ui/packages/ui-core/src/workspace/WorkspaceRoot.tsx:52]: I went through them all and sensors was the only one that had an issue which I fixed

File: js_modules/dagster-ui/packages/ui-core/src/workspace/WorkspaceContext.tsx (hunk 5 above)
Comment (salazarm) [reply js_modules/dagster-ui/packages/ui-core/src/workspace/WorkspaceContext.tsx:199]: I'll do this in a follow up pr, I think we just want to do this generally in the app across localStorage and indexedDB both.

File: js_modules/dagster-ui/packages/ui-core/src/search/useIndexedDBCachedQuery.tsx [hunk 8]
```diff
@@ -84,9 +84,16 @@ export function useIndexedDBCachedQuery<TQuery, TVariables extends OperationVari
...
-    fetchState[key]?.onFetched.forEach((cb) => cb(queryResult));
     delete fetchState[key];
+    const onFetched = fetchState[key]?.onFetched;
+    try {
+      setData(data);
+    } catch (e) {}
+    onFetched?.forEach((cb) => {
+      try {
+        cb(queryResult);
+      } catch (e) {}
```
Comment (bengotow) [inline js_modules/dagster-ui/packages/ui-core/src/search/useIndexedDBCachedQuery.tsx:95]: Just calling out that this mutes a potentially significant amount of code since lots of stuff could run downstream of the data arriving - this should probably still emit a datadog trace + console.error just in case!

**Review by bengotow — APPROVED** [review]

File: js_modules/dagster-ui/packages/ui-core/src/search/useIndexedDBCachedQuery.tsx (hunk 8 above)
Comment (salazarm) [reply js_modules/dagster-ui/packages/ui-core/src/search/useIndexedDBCachedQuery.tsx:95]: Ah yeah I'll rethrow the error in a setTimeout. We need a general utility for this.
